test/resolver_test.go

```diff
@@ -0,0 +1,215 @@
+			gotContextChan.SendContext(ctx, ctx)
+			return &testpb.Empty{}, nil
+		},
+	}
+	ss.r = manual.NewBuilderWithScheme("confSel")
+
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctxDeadline := time.Now().Add(10 * time.Second)
+	ctx, cancel := context.WithDeadline(context.Background(), ctxDeadline)
+	defer cancel()
+
+	state := iresolver.SetConfigSelector(resolver.State{
+		Addresses:     []resolver.Address{{Addr: ss.address}},
+		ServiceConfig: parseCfg(ss.r, "{}"),
+	}, funcConfigSelector{
+		f: func(i iresolver.RPCInfo) *iresolver.RPCConfig {
+			gotInfoChan.Send(&i)
+			cfgI, err := sendConfigChan.Receive(ctx)
+			if err != nil {
+				t.Errorf("error waiting for RPCConfig: %v", err)
+				return nil
+			}
+			cfg := cfgI.(*iresolver.RPCConfig)
+			if cfg != nil && cfg.Context == nil {
+				cfg.Context = i.Context
+			}
+			return cfg
+		},
+	})
+	ss.r.UpdateState(state) // Blocks until config selector is applied
+
+	longCtxDeadline := time.Now().Add(30 * time.Second)
+	longdeadlineCtx, cancel := context.WithDeadline(context.Background(), longCtxDeadline)
+	defer cancel()
+	shorterTimeout := 3 * time.Second
+
+	testMD := metadata.MD{"footest": []string{"bazbar"}}
+	mdOut := metadata.MD{"handler": []string{"value"}}
+
+	var onCommittedCalled bool
+
+	testCases := []struct {
+		name   string
+		md     metadata.MD
+		config *iresolver.RPCConfig
+
+		wantMD       metadata.MD
+		wantDeadline time.Time
+		wantTimeout  time.Duration
+	}{{
+		name:         "basic",
+		md:           testMD,
+		config:       &iresolver.RPCConfig{},
+		wantMD:       testMD,
+		wantDeadline: ctxDeadline,
+	}, {
+		name: "alter MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: metadata.NewOutgoingContext(ctx, mdOut),
+		},
+		wantMD:       mdOut,
+		wantDeadline: ctxDeadline,
+	}, {
+		name: "alter timeout; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: longdeadlineCtx, // no metadata
+		},
+		wantMD:       nil,
+		wantDeadline: longCtxDeadline,
+	}, {
+		name:         "nil config",
+		md:           metadata.MD{},
+		config:       nil,
+		wantMD:       nil,
+		wantDeadline: ctxDeadline,
+	}, {
+		name: "alter timeout via method config; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			MethodConfig: serviceconfig.MethodConfig{
+				Timeout: &shorterTimeout,
+			},
+		},
+		wantMD:      nil,
+		wantTimeout: shorterTimeout,
+	}, {
+		name: "onCommitted callback",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			OnCommitted: func() {
+				onCommittedCalled = true
+			},
+		},
+		wantMD:       testMD,
+		wantDeadline: ctxDeadline,
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if !sendConfigChan.SendOrFail(tc.config) {
+				t.Fatalf("last config not consumed by config selector")
+			}
+
+			onCommittedCalled = false
+			ctx = metadata.NewOutgoingContext(ctx, tc.md)
```

Comment:
Is this not `ctx := metadata.NewOutgoingContext(ctx, tc.md)` by choice?

Reply:
No, that's a mistake.  Fixed.